Add execution context stack for accessing caller information

aviralg opened the pull request:
This commit adds an execution context stack to the tracer.
Every time a closure, builtin or special is entered, a
context is pushed on the stack and on exiting from the function
the context is popped.

This provides access to the entire call chain. This can be used
to get information such as the caller of the currently entered
function or the parent environment.

eval expression analysis has also been updated to utilize the
stack and get the caller of the eval function.

aviralg committed:
Add execution context stack for accessing caller information

This commit adds an execution context stack to the tracer.
Every time a closure, builtin or special is entered, a
context is pushed on the stack and on exiting from the function
the context is popped.

This provides access to the entire call chain. This can be used
to get information such as the caller of the currently entered
function or the parent environment.

eval expression analysis has also been updated to utilize the
stack and get the caller of the eval function.

aviralg commented:
There is still some polishing needed. Overall this is quite close to being done.
- Currently the stack is not updated on context jumps inside the interpreter. I will address this next.
- `EvalExpressionAnalysis` tries to find the caller of `Eval`. It finds the immediate parent of `eval` from the call stack and this does not necessarily provide what we want. For example -
```
f <- function() {
    x <- eval(quote(1 + 2))
}
f()
```
In this code example, we expect the caller of `eval` to be `f`. But the immediate parent on the call stack will be `<-`. This "smarts" has to be added to the `EvalExpressionAnalysis`.

aviralg committed:
Add support for stack unwinding on context jumps

This commit tracks context jumps inside the R interpreter
and uses it to update the execution stack.
This ensures that entries on stack correctly represent
currently executing functions.

The ExecutionContext classes have been created to
reflect their hierarchy and purpose.

EvalExpressionAnalysis has been updated to provide
information about the closure that calls eval.

aviralg commented:
Fixed both issues.

fikovnik commented on src/AnalysisDriver.cpp:
```diff
@@ -1,10 +1,91 @@
+      execution_context_stack_{std::make_shared<ExecutionContextStack>()},
+      eval_expression_analysis_{configuration, execution_context_stack_} {}
+
+void AnalysisDriver::dyntrace_entry(const SEXP expression,
```
This is minor, but I think both params will fit on one line

aviralg replied:
clang-format :-)

fikovnik commented on src/AnalysisDriver.cpp:
```diff
@@ -1,10 +1,91 @@
+    execution_context_stack_->push(context);
+}
+
+void AnalysisDriver::dyntrace_exit(const SEXP expression,
```
Same here? Or even distribution?

aviralg replied:
clang-format :-)

fikovnik commented on src/AnalysisDriver.cpp:
```diff
@@ -1,10 +1,91 @@
+                                   const int error) {
+
+    TopLevelExecutionContext context{
+        std::get<TopLevelExecutionContext>(execution_context_stack_->pop())};
```
Can we put the `}` on the new line?

aviralg replied:
Its auto-formatted. We need to configure clang-format to do it differently.

fikovnik commented on src/AnalysisDriver.cpp:
```diff
@@ -1,10 +1,91 @@
+                                  const SEXP args, const SEXP rho,
+                                  const SEXP retval) {
+    SpecialExecutionContext context{
+        std::get<SpecialExecutionContext>(execution_context_stack_->pop())};
```
same here

aviralg replied:
Its auto-formatted. We need to configure clang-format to do it differently.

fikovnik commented on src/AnalysisDriver.cpp:
```diff
@@ -1,10 +1,91 @@
+                                  const SEXP retval) {
+
+    BuiltinExecutionContext context{
+        std::get<BuiltinExecutionContext>(execution_context_stack_->pop())};
```
same here

aviralg replied:
Its auto-formatted. We need to configure clang-format to do it differently.

fikovnik commented on src/AnalysisDriver.cpp:
```diff
@@ -1,10 +1,91 @@
+void AnalysisDriver::context_jump(const RCNTXT *context,
+                                  const SEXP return_value, int restart) {
+    execution_contexts_t unwound_contexts{
+        execution_context_stack_->unwind(RExecutionContext{context})};
```
same here

fikovnik commented on src/AnalysisDriver.cpp:
```diff
@@ -1,10 +1,91 @@
+
+void AnalysisDriver::context_exit(const RCNTXT *rcontext) {
+    RExecutionContext context{
+        std::get<RExecutionContext>(execution_context_stack_->pop())};
```
same here

aviralg replied:
Its auto-formatted. We need to configure clang-format to do it differently.

fikovnik commented on src/ClosureExecutionContext.h:
```diff
@@ -0,0 +1,28 @@
+        return (get_call() == rhs.get_call() &&
+                get_function() == rhs.get_function() &&
+                get_arguments() == rhs.get_arguments() &&
+                get_environment() == rhs.get_environment());
```
and here
```
return (
 ...
);
```

aviralg replied:
Its auto-formatted. We need to configure clang-format to do it differently.

fikovnik commented on src/ExecutionContext.h:
```diff
@@ -0,0 +1,18 @@
+
+class ExecutionContext {
+  public:
+    explicit ExecutionContext() {}
```
Do we need explicit here?

aviralg replied:
Its a rule of thumb to always make constructors `explicit` unless you want implicit type conversion to objects of this type.

fikovnik commented on src/ExecutionContextStack.h:
```diff
@@ -0,0 +1,92 @@
+            execution_context_t temp_context{pop()};
+
+            if (std::holds_alternative<RExecutionContext>(temp_context)) {
+
```
extra new line?

fikovnik commented on src/ExecutionContextStack.h:
```diff
@@ -0,0 +1,92 @@
+        dyntrace_log_error("cannot find matching execution context\n");
+    }
+
+    std::optional<execution_context_t>
```
extra new line?

aviralg committed:
Add source location and dynamic function call count to analysis

This commit adds information about the caller of the eval function
such as line number, column number, directory and filename.

The number of closures, specials and builtins called while eval
executes is also tracked and reported.

aviralg committed:
Update source location computation and fix include guards
